### PR TITLE
Add fn install & fn uninstall command

### LINKED ISSUE #4: Optimized installation approach
We need to improve the way OpenFunction is installed and deployed.

Integrating them into cli project would be a better solution.

### PR DESCRIPTION
**Pull request by tpiperatgod**
Add fn install & fn uninstall command

Signed-off-by: laminar <fangtian@kubesphere.io>

### CODE REVIEW

**Comment by tpiperatgod** [comment]
Here are two examples of `install` and `uninstall`:

### fn install
```shell
# fn install --all --upgrade
Start installing OpenFunction and its dependencies.
Here are the components and corresponding versions to be installed for this installation:
+------------------+---------+
| COMPONENT        | VERSION |
+------------------+---------+
| Keda             | 2.4.0   |
| Knative Serving  | 0.26.0  |
| Cert Manager     | v1.5.4  |
| Ingress Nginx    | 1.1.0   |
| Tekton Pipelines | v0.28.1 |
| Shipwright       | 0.6.0   |
| OpenFunction     | v0.4.0  |
| Dapr             | 1.4.3   |
+------------------+---------+
The following components already exist in the current cluster:
+------------------+---------+
| COMPONENT        | VERSION |
+------------------+---------+
| Keda             |         |
| Knative Serving  |         |
| Ingress Nginx    |         |
| Cert Manager     |         |
| Tekton Pipelines |         |
| Dapr             |         |
+------------------+---------+
You can see the list of components to be installed and the list of components already exist in the cluster.
You have used the `--upgrade` parameter, which means that the installation process will overwrite the components that already exist in the cluster.
Make sure you know what happens when you do this.
Enter 'y' to continue and 'n' to abort:
-> y
🔄  -> INGRESS <- Installing Ingress...
🔄  -> KNATIVE <- Installing Knative Serving...
🔄  -> DAPR <- Installing Dapr...
🔄  -> DAPR <- Downloading Dapr Cli binary...
🔄  -> KEDA <- Installing Keda...
🔄  -> CERTMANAGER <- Installing Cert Manager...
🔄  -> SHIPWRIGHT <- Installing Shipwright...
🔄  -> INGRESS <- Checking if Ingress is ready...
🔄  -> KEDA <- Checking if Keda is ready...
🔄  -> CERTMANAGER <- Checking if Cert Manager is ready...
🔄  -> SHIPWRIGHT <- Checking if Shipwright is ready...
🔄  -> KNATIVE <- Installing Kourier as Knative's gateway...
🔄  -> KNATIVE <- Configuring Knative Serving's DNS...
🔄  -> KNATIVE <- Checking if Knative Serving is ready...
✅  -> CERTMANAGER <- Done!
🔄  -> DAPR <- Initializing Dapr with Kubernetes mode...
✅  -> SHIPWRIGHT <- Done!
✅  -> KNATIVE <- Done!
✅  -> INGRESS <- Done!
✅  -> DAPR <- Done!
✅  -> KEDA <- Done!
🔄  -> OPENFUNCTION <- Installing OpenFunction...
🔄  -> OPENFUNCTION <- Checking if OpenFunction is ready...
✅  -> OPENFUNCTION <- Done!
🚀 Completed in 2m3.638035129s.
```

### fn uninstall
```shell
~# fn uninstall --all
Start uninstalling OpenFunction and its dependencies.
Here are the components and corresponding versions to be uninstalled for this installation:
+------------------+---------+
| COMPONENT        | VERSION |
+------------------+---------+
| Cert Manager     | v1.5.4  |
| Ingress Nginx    | 1.1.0   |
| Tekton Pipelines | v0.28.1 |
| Shipwright       | 0.6.0   |
| OpenFunction     | v0.4.0  |
| Dapr             | 1.4.3   |
| Keda             | 2.4.0   |
| Knative Serving  | 0.26.0  |
+------------------+---------+
You can see the list of components to be installed and the list of components already exist in the cluster.
You have used the `--upgrade` parameter, which means that the installation process will overwrite the components that already exist in the cluster.
Make sure you know what happens when you do this.
Enter 'y' to continue and 'n' to abort:
-> y
🔄  -> OPENFUNCTION <- Uninstalling OpenFunction...
🔄  -> KNATIVE <- Uninstalling Knative Serving...
🔄  -> DAPR <- Uninstalling Dapr with Kubernetes mode...
🔄  -> KEDA <- Uninstalling Keda...
🔄  -> SHIPWRIGHT <- Uninstalling Tekton Pipeline & Shipwright...
🔄  -> INGRESS <- Uninstalling Ingress...
🔄  -> CERTMANAGER <- Uninstalling Cert Manager...
✅  -> OPENFUNCTION <- Done!
✅  -> DAPR <- Done!
🔄  -> KNATIVE <- Uninstalling Kourier...
✅  -> KEDA <- Done!
✅  -> CERTMANAGER <- Done!
✅  -> KNATIVE <- Done!
✅  -> INGRESS <- Done!
✅  -> SHIPWRIGHT <- Done!
🚀 Completed in 1m21.683329262s.
```

**Comment by tpiperatgod** [comment]
And provide `make build` command to build `fn` binary. The `fn` binary can be found in the `./dist` directory.

```shell
~# make build
go fmt ./...
/opt/openfunction/fn-cli/bin/goimports -w cmd/ pkg/ testdata/
go vet ./...
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build  -ldflags "-s -w -X 'main.goversion=go version go1.16.7 linux/amd64'" \
-o ./dist/fn_linux_amd64 cmd/main.go;
```

File: pkg/dependency/linux/linux.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	}
+	return nil
+}
+
+func (e *Executor) GetExistDaprVerion() string {
+	cmd := "kubectl get namespace dapr-system"
+	if _, _, err := e.Exec(cmd); err != nil {
+		return "None"
+	} else {
+		cmd = "kubectl get po -n dapr-system -L app.kubernetes.io/version|grep operator|awk '{ print $6 }'"
```
Comment (wanjunlei) [inline pkg/dependency/linux/linux.go]: The return value of my execution of this command is `ago)`, and the following command can return the correct result.

```
kubectl get po -n dapr-system -L app.kubernetes.io/version|grep operator|awk '{ print $8 }'
```

The K8s version is `1.22.1`, does it related to the K8s version?

**Review by wanjunlei — CHANGES_REQUESTED** [review]

File: pkg/dependency/linux/linux.go (hunk 1 above)
Comment (tpiperatgod) [reply pkg/dependency/linux/linux.go]: This has been deprecated and I will delete it.

Now I use the clientset to get the version of components:

```go
func GetExistComponentVersion(ctx context.Context, cl *k8s.Clientset, ns string, deploymentName string) string {
	if deploy, err := cl.AppsV1().Deployments(ns).Get(ctx, deploymentName, metav1.GetOptions{}); err != nil {
		return ""
	} else {
		var labels map[string]string
		if ns == IngressNginxNamespace || ns == KedaNamespace {
			labels = deploy.GetLabels()
		} else {
			labels = deploy.Spec.Template.GetLabels()
		}
		if version, ok := labels[kubernetestVersionLabel]; !ok {
			return ""
		} else {
			return version
		}
	}
}
```

Comment (tpiperatgod) [reply pkg/dependency/linux/linux.go]: So I think we might suggest shipwright add the label like "app.kubernetes.io/version" as well.

Comment (wanjunlei) [reply pkg/dependency/linux/linux.go]: Yes, I will add these labels later.

**Review by wanjunlei — APPROVED** [review]

File: docs/install.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+--async            For installing OpenFunction Async Runtime(Dapr & Keda).
+--cert-manager     For installing Cert Manager.
+--dapr             For installing Dapr.
+--dry-run          Used to prompt for the components and their versions to be installed by the current command.
+--ingress          For installing Ingress Nginx.
+--keda             For installing Keda.
+--knative          For installing Knative Serving (with Kourier as default gateway).
+--region-cn        For users in China to speed up the download process of dependent components.
+--shipwright       For installing ShipWright.
+--sync             For installing OpenFunction Sync Runtime(Knative).
```
Comment (benjaminhuo) [inline docs/install.md]: --sync means OpenFunctionSync runtime and probably needs to install [keda-http-addon](https://github.com/kedacore/http-add-on/blob/main/docs/install.md)  in the future, --knative is a seperate runtime

File: docs/install.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+or
+
+```shell
+fn install --sync
+```
+
+#### Support for users in China to speed up the installation process
+
+> When you uninstall, you must also add `--region-cn`.
```
Comment (benjaminhuo) [inline docs/install.md]: when you uninstall? the cmd below is install

File: pkg/cmd/subcommand/install.go [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+
+	// Update the corresponding conditions when WithAsyncRuntime is true
+	if i.WithAsyncRuntime {
+		i.WithDapr = true
+		i.WithKeda = true
+	}
+
+	// Update the corresponding conditions when WithSyncRuntime is true
+	if i.WithSyncRuntime {
+		i.WithKnative = true
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/install.go]: Can we use SyncRuntime for keda-http + deployment combination only, aka OpenFunctionSync runtime.
knative could be a separate option although it's now categorized into sync runtime too

Comment (tpiperatgod) [reply pkg/cmd/subcommand/install.go]: Understood, I've commented it for now

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**